From: Frank van der Linden <fvdl@google.com>
Subject: mm: make SPARSEMEM_VMEMMAP_PREINIT an internal option
Date: Thu, 27 Feb 2025 18:57:17 +0000

SPARSEMEM_VMEMMAP_PREINIT is not useful as a user-facing option, it is
just something that should be selected if a subsystem wishes to do vmemmap
preinit.  That's currently only HUGETLB_PAGE_OPTIMIZE_VMEMMAP.

So, make it a default-less option that is only selected by
HUGETLB_PAGE_OPTIMIZE_VMEMMAP, iff the architecture has noted it is
capable of doing hugetlb vmemmap preinit.  That is done via
ARCH_WANT_HUGETLB_VMEMMAP_PREINIT, renamed from
ARCH_WANT_SPARSE_VMEMMAP_PREINIT.

Link: https://lkml.kernel.org/r/20250227185717.1265644-1-fvdl@google.com
Fixes: a916062fdd134 ("mm/sparse: allow for alternate vmemmap section init at boot")
Signed-off-by: Frank van der Linden <fvdl@google.com>
Acked-by: Johannes Weiner <hannes@cmpxchg.org>
Signed-off-by: Andrew Morton <akpm@linux-foundation.org>
---

 arch/x86/Kconfig |    1 +
 fs/Kconfig       |    1 +
 mm/Kconfig       |    9 ++++-----
 3 files changed, 6 insertions(+), 5 deletions(-)

--- a/arch/x86/Kconfig~mm-sparse-allow-for-alternate-vmemmap-section-init-at-boot-fix
+++ a/arch/x86/Kconfig
@@ -146,6 +146,7 @@ config X86
 	select ARCH_WANT_LD_ORPHAN_WARN
 	select ARCH_WANT_OPTIMIZE_DAX_VMEMMAP	if X86_64
 	select ARCH_WANT_OPTIMIZE_HUGETLB_VMEMMAP	if X86_64
+	select ARCH_WANT_HUGETLB_VMEMMAP_PREINIT if X86_64
 	select ARCH_WANTS_THP_SWAP		if X86_64
 	select ARCH_HAS_PARANOID_L1D_FLUSH
 	select BUILDTIME_TABLE_SORT
--- a/fs/Kconfig~mm-sparse-allow-for-alternate-vmemmap-section-init-at-boot-fix
+++ a/fs/Kconfig
@@ -286,6 +286,7 @@ config HUGETLB_PAGE_OPTIMIZE_VMEMMAP
 	def_bool HUGETLB_PAGE
 	depends on ARCH_WANT_OPTIMIZE_HUGETLB_VMEMMAP
 	depends on SPARSEMEM_VMEMMAP
+	select SPARSEMEM_VMEMMAP_PREINIT if ARCH_WANT_HUGETLB_VMEMMAP_PREINIT
 
 config HUGETLB_PMD_PAGE_TABLE_SHARING
 	def_bool HUGETLB_PAGE
--- a/mm/Kconfig~mm-sparse-allow-for-alternate-vmemmap-section-init-at-boot-fix
+++ a/mm/Kconfig
@@ -443,13 +443,9 @@ config SPARSEMEM_VMEMMAP
 	  pfn_to_page and page_to_pfn operations.  This is the most
 	  efficient option when sufficient kernel resources are available.
 
-config ARCH_WANT_SPARSEMEM_VMEMMAP_PREINIT
+config SPARSEMEM_VMEMMAP_PREINIT
 	bool
 
-config SPARSEMEM_VMEMMAP_PREINIT
-	bool "Early init of sparse memory virtual memmap"
-	depends on SPARSEMEM_VMEMMAP && ARCH_WANT_SPARSEMEM_VMEMMAP_PREINIT
-	default y
 #
 # Select this config option from the architecture Kconfig, if it is preferred
 # to enable the feature of HugeTLB/dev_dax vmemmap optimization.
@@ -460,6 +456,9 @@ config ARCH_WANT_OPTIMIZE_DAX_VMEMMAP
 config ARCH_WANT_OPTIMIZE_HUGETLB_VMEMMAP
 	bool
 
+config ARCH_WANT_HUGETLB_VMEMMAP_PREINIT
+	bool
+
 config HAVE_MEMBLOCK_PHYS_MAP
 	bool
 
_
